pkg/kubelet/volume/nfs/nfs.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nfs
+
+import (
+	"os"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/mount"
+	"github.com/golang/glog"
+)
+
+// This is the primary entrypoint for volume plugins.
+func ProbeVolumePlugins() []volume.Plugin {
+	return []volume.Plugin{&nfsPlugin{nil}}
+}
+
+type nfsPlugin struct {
+	host volume.Host
+}
+
+var _ volume.Plugin = &nfsPlugin{}
+
+const (
+	nfsPluginName = "kubernetes.io/nfs"
+)
+
+func (plugin *nfsPlugin) Init(host volume.Host) {
+	plugin.host = host
+}
+
+func (plugin *nfsPlugin) Name() string {
+	return nfsPluginName
+}
+
+func (plugin *nfsPlugin) CanSupport(spec *api.Volume) bool {
+	if spec.VolumeSource.NFS != nil {
+		return true
+	}
+	return false
+}
+
+func (plugin *nfsPlugin) NewBuilder(spec *api.Volume, podUID types.UID) (volume.Builder, error) {
+	return plugin.newBuilderInternal(spec, podUID, NewNFSMounter())
+}
+
+func (plugin *nfsPlugin) newBuilderInternal(spec *api.Volume, podUID types.UID, mounter NFSMountInterface) (volume.Builder, error) {
+	return &nfs{
+		volName:      spec.Name,
+		server:       spec.VolumeSource.NFS.Server,
+		exportPath:   spec.VolumeSource.NFS.SourcePath,
+		mountOptions: spec.VolumeSource.NFS.MountOptions,
+		mounter:      mounter,
+		podUID:       podUID,
+		plugin:       plugin,
+	}, nil
+}
+
+func (plugin *nfsPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID, NewNFSMounter())
+}
+
+func (plugin *nfsPlugin) newCleanerInternal(volName string, podUID types.UID, mounter NFSMountInterface) (volume.Cleaner, error) {
+	return &nfs{
+		volName:      volName,
+		server:       "",
+		exportPath:   "",
+		mountOptions: "",
+		mounter:      mounter,
+		podUID:       podUID,
+		plugin:       plugin,
+	}, nil
+}
+
+// NFS volumes represent a bare host file or directory mount of an NFS export.
+type nfs struct {
+	volName      string
+	podUID       types.UID
+	server       string
+	exportPath   string
+	mountOptions string
+	mounter      NFSMountInterface
+	plugin       *nfsPlugin
+}
+
+func (nfs *nfs) SetUp() error {
+	mountpoint, err := mount.IsMountPoint(nfs.GetPath())
```

Comment:
I was actually discussing this with @rootfs and another RH developer on Friday. 

RH will ultimately have several more volumes to contribute besides NFS and iSCSCI.   I think we'll be in a good position to see all the commonalities and make a sensible whole from them.

By the time we introduce the 3rd new volume, we'll have strongly recognizable patterns in place and can refactor all to use common code.  This also gives us 2 new working plugins in the meantime.